Zend/zend_inheritance.c

```diff
@@ -1110,17 +1110,38 @@ static void zend_do_inherit_interfaces(zend_class_entry *ce, const zend_class_en
 }
 /* }}} */
 
+inheritance_status class_constant_types_compatible(const zend_class_constant *parent, const zend_class_constant *child) /* {{{ */
+{
+	return zend_perform_covariant_type_check(child->ce, child->type, parent->ce, parent->type);
+}
+/* }}} */
+
 static void do_inherit_class_constant(zend_string *name, zend_class_constant *parent_const, zend_class_entry *ce) /* {{{ */
 {
 	zval *zv = zend_hash_find_ex(&ce->constants_table, name, 1);
 	zend_class_constant *c;
 
 	if (zv != NULL) {
 		c = (zend_class_constant*)Z_PTR_P(zv);
+
 		if (UNEXPECTED((Z_ACCESS_FLAGS(c->value) & ZEND_ACC_PPP_MASK) > (Z_ACCESS_FLAGS(parent_const->value) & ZEND_ACC_PPP_MASK))) {
 			zend_error_noreturn(E_COMPILE_ERROR, "Access level to %s::%s must be %s (as in class %s)%s",
 				ZSTR_VAL(ce->name), ZSTR_VAL(name), zend_visibility_string(Z_ACCESS_FLAGS(parent_const->value)), ZSTR_VAL(parent_const->ce->name), (Z_ACCESS_FLAGS(parent_const->value) & ZEND_ACC_PUBLIC) ? "" : " or weaker");
 		}
+
+		if (!(Z_ACCESS_FLAGS(parent_const->value) & ZEND_ACC_PRIVATE)) {
+			if (UNEXPECTED(ZEND_TYPE_IS_SET(parent_const->type))) {
+				if (class_constant_types_compatible(parent_const, c) == INHERITANCE_ERROR) {
+					zend_string *type_str = zend_type_to_string_resolved(parent_const->type, parent_const->ce);
+
+					zend_error_noreturn(E_COMPILE_ERROR, "Type of %s::%s must be %s (as in class %s)",
```

Comment:
Error message is inaccurate. Sub-types will work too, no? How about "Declaration of ... must be compatible with ..."

Reply:
I matched the error message with typed properties for consistency.

Reply:
Yeah but properties are not covariant since you can read from and write to them.

Reply:
> Yeah but properties are not covariant since you can read from and write to them.

Oh, I see what you mean. Good catch!